### ISSUE TITLE
Docker image is not working on Apple M1/M2 chips

### ISSUE THREAD

**Issue by mvacher** [zama-ai/concrete-ml#160]
## Summary

The version provided through the docker image is not working on Apple M1 and M2 chips

## Description

<details><summary>minimal POC to trigger the bug</summary>
<p>

On a M1 or M2 machine:

```
# Install
docker pull zamafhe/concrete-ml:latest

# Run the with:
docker run --rm -it -p 8888:8888 -v ~/host_folder:/data zamafhe/concrete-ml
```

Python code: ( example from the readme)

```
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from concrete.ml.sklearn import LogisticRegression


# Lets create a synthetic data-set
x, y = make_classification(n_samples=100, class_sep=2, n_features=30, random_state=42)

# Split the data-set into a train and test set
X_train, X_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42
)
# Now we train in the clear and quantize the weights
model = LogisticRegression(n_bits=8)
model.fit(X_train, y_train)

# We then compile on a representative set
model.compile(X_train) # --> This line kills the kernel and restart the notebook
```

</p>
</details>

**Comment by RomanBredehoft** [comment]
We indeed don't handle docker with M1 / M2 chips currently but we'll try our best to patch it soon (I'm on M2 as well 😉). Thanks for the issue ! In the mean time, please use the native installation as Concrete-ML properly works with Apple Silicon !

**Comment by bcm-at-zama** [comment]
Hello @mvacher. We're having a look, sorry for the delay. Were you able to use Concrete with native packages?

**Comment by berhankarabay** [comment]
Hello, I'm on M1 and I've tried several options. Unfortunately I wasn't able to run Concrete-ML. It gives 'Illegal Instruction' in Kernel. Can you guide me through using Concrete-ML with native packages?

**Comment by bcm-at-zama** [comment]
Hello @berhankarabay . Did you try to install natively, ie with

```
pip install -U pip wheel setuptools
pip install concrete-ml
```

ideally in a virtual env?

**Comment by berhankarabay** [comment]
Hello,

Thank you for your quick return. I'm giving the details below.

I'm using Spyder through Anaconda. At first, I created a new environment in
Anaconda. Then, I completed the installation using the pip commands as
suggested. All works fine.

Unfortunately, when I tried to run my code, I still got the following
error. As you can see in the screenshot, it gives 'Fatal Python error:
Illegal Instruction' and restarts the kernel. I couldn't understand why it
happens. I took this part of the code from a similar work
<https://www.kdnuggets.com/2022/08/machine-learning-encrypted-data.html#:~:text=This%20was%20possible%20thanks%20to,Machine%20Learning%20Engineer%20at%20Zama>
using
a machine with 8 11th Gen Intel® Core™ i5-1135G7 processors of 2.40GHz.

[image: Screenshot 2023-11-16 at 13.26.05.png]

To see if any other application would work, I tried the example usage given
in your website <https://docs.zama.ai/concrete-ml>, and I got this
following error. It doesn't recognize the fhe argument.
[image: Screenshot 2023-11-16 at 13.35.42.png]

Can you help me with this? I'm doing my master's thesis and I really want
to use concrete-ml for it.

Best regards,

**Comment by fd0r** [comment]
Hello @berhankarabay , are you using a M1 system or intel based system?
Also what snippet of code gave you this error?
Also which version of python are you using in your conda env?
Best regards

**Comment by bcm-at-zama** [comment]
Yes, you might want to start with the very small example which is given in https://github.com/zama-ai/concrete-ml#a-simple-concrete-ml-example-with-scikit-learn to test your config

**Comment by berhankarabay** [comment]
I shared again in the attachment. Can you see it? Thanks.

**Comment by bcm-at-zama** [comment]
Sorry we don't see attachments, maybe it doesn't work by email? Maybe just copy-paste the text of the stdio, if you can't make it work

**Comment by fd0r** [comment]
Alright so 3.8 should work nicely.
Could you please replace `execute_in_fhe=True` by `fhe="execute"` ?

**Comment by berhankarabay** [comment]
Luis, I've tried it but I've received the same error with the example
usage. Sharing the script below.

fhe_predictions = concrete_model.best_estimator_.predict(

  rescaledValidationX, fhe="execute"
)

TypeError: predict() got an unexpected keyword argument 'fhe'

Benoit, I'm also sharing the previous code and error script below.

This is from example usage: y_pred_fhe = model.predict(X_test,
fhe="execute")

TypeError: predict() got an unexpected keyword argument 'fhe'

the_predictions = concrete_model.best_estimator_.predict(

  rescaledValidationX, execute_in_fhe=True
)

Fatal Python error: Illegal instruction


Main thread:
Current thread 0x00000002008e1240 (most recent call first):
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/compiler/client_support.py",
line 84 in key_set
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/numpy/compilation/client.py",
line 106 in keygen
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/numpy/compilation/client.py",
line 173 in encrypt
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/numpy/compilation/circuit.py",
line 130 in encrypt
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/numpy/compilation/circuit.py",
line 188 in encrypt_run_decrypt
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/ml/sklearn/base.py",
line 690 in _execute_in_fhe
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/ml/sklearn/base.py",
line 596 in predict_proba
  File
"/Users/berhankarabay/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/ml/sklearn/base.py",
line 573 in predict
  File
"/var/folders/gp/s1b5k9yn3ks7tc9zsq5xfs4w0000gn/T/ipykernel_62348/1350764500.py",
line 1 in <module>


Restarting kernel...

**Comment by berhankarabay** [comment]
Hello,

I've also tried using Jupyter Notebook instead of Spyder. To avoid any
complexity, I tried to run the example usage
<https://docs.zama.ai/concrete-ml> from your website.

Before the execution of fhe, I've encountered the following error. Can you
check it? It may be related to the other issue as well.

RuntimeError                              Traceback (most recent call last)
Cell In[1], line 21     18 y_pred_clear = model.predict(X_test)     20
# We then compile on a representative set---> 21
model.compile(X_train)     23 # Finally we run the inference on
encrypted inputs     24 y_pred_fhe = model.predict(X_test,
fhe="execute")

File ~/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/ml/sklearn/base.py:1101,
in SklearnLinearModelMixin.compile(self, X, configuration,
compilation_artifacts, show_mlir, use_virtual_lib, p_error)   1098
quantized_numpy_inputset = self.quantized_module_.quantize_input(X)
1100 # Compile the model-> 1101 circuit =
self.quantized_module_.compile(   1102     quantized_numpy_inputset,
1103     configuration,   1104     compilation_artifacts,   1105
show_mlir=show_mlir,   1106     use_virtual_lib=use_virtual_lib,
1107     p_error=p_error,   1108 )   1110 return circuit

File ~/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/ml/quantization/quantized_module.py:358,
in QuantizedModule.compile(self, q_inputs, configuration,
compilation_artifacts, show_mlir, use_virtual_lib, p_error)    354
return (numpy.expand_dims(arr, 0) for arr in q_inputs[0])    356
inputset = get_inputset_iterable()--> 358 self.forward_fhe =
compiler.compile(    359     inputset,    360     configuration,
361     compilation_artifacts,    362     show_mlir=show_mlir,    363
   virtual=use_virtual_lib,    364     p_error=p_error,    365 )
367 self._is_compiled = True    369 return self.forward_fhe

File ~/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/numpy/compilation/compiler.py:373,
in Compiler.compile(self, inputset, configuration, artifacts,
**kwargs)    370 self._evaluate("Compiling", inputset)    371 assert
self.graph is not None--> 373 mlir =
GraphConverter.convert(self.graph, virtual=self.configuration.virtual)
   374 if self.artifacts is not None:    375
self.artifacts.add_mlir_to_compile(mlir)

File ~/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/numpy/mlir/graph_converter.py:401,
in GraphConverter.convert(graph, virtual)    398 if virtual:    399
 return "Virtual circuits doesn't have MLIR."--> 401
GraphConverter._update_bit_widths(graph)    402
GraphConverter._offset_negative_lookup_table_inputs(graph)    404 #
There are no tensor +*- scalar operations in the compiler    405 # But
such operations are used commonly, so we need to support them    406 #
So, we implemented some workarounds (pull request #970)   (...)    409
    410 # { node1: "%arg0", node2: "%0", node3: "%1" }

File ~/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/concrete/numpy/mlir/graph_converter.py:226,
in GraphConverter._update_bit_widths(graph)    221
offending_nodes[node] = [    222             f"only up to
{MAXIMUM_BIT_WIDTH}-bit integers are supported"    223         ]
225 if len(offending_nodes) != 0:--> 226     raise RuntimeError(
227         "Function you are trying to compile cannot be converted to
MLIR:\n\n"    228         +
graph.format(highlighted_nodes=offending_nodes)    229     )    231
for node in graph.graph.nodes:    232     for value in node.inputs +
[node.output]:
RuntimeError: Function you are trying to compile cannot be converted to MLIR:

%0 = [[ -3] [-1 ... -1] [  4]]        # ClearTensor<int8, shape=(30, 1)>
%1 = _input_0                         # EncryptedTensor<uint8, shape=(1, 30)>
%2 = -131                             # ClearScalar<int9>
%3 = add(%1, %2)                      # EncryptedTensor<int9, shape=(1, 30)>
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
only up to 8-bit integers are supported
%4 = subgraph(%3)                     # EncryptedTensor<int8, shape=(1, 30)>
%5 = matmul(%4, %0)                   # EncryptedTensor<int16, shape=(1, 1)>
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
only up to 8-bit integers are supported
%6 = subgraph(%5)                     # EncryptedTensor<uint8, shape=(1, 1)>
return %6

Subgraphs:

    %4 = subgraph(%3):

         %0 = -128                           # ClearScalar<int8>
         %1 = 127                            # ClearScalar<uint7>
         %2 = 3                              # ClearScalar<uint2>
         %3 = 0.02919340787200249            # ClearScalar<float64>
         %4 = 0.02919340787200249            # ClearScalar<float64>
         %5 = input                          # EncryptedTensor<uint2,
shape=(1, 30)>
         %6 = multiply(%4, %5)               #
EncryptedTensor<float64, shape=(1, 30)>
         %7 = divide(%6, %3)                 #
EncryptedTensor<float64, shape=(1, 30)>
         %8 = add(%7, %2)                    #
EncryptedTensor<float64, shape=(1, 30)>
         %9 = rint(%8)                       #
EncryptedTensor<float64, shape=(1, 30)>
        %10 = clip(%9, %0, %1)               #
EncryptedTensor<float64, shape=(1, 30)>
        %11 = astype(%10, dtype=int_)        # EncryptedTensor<uint1,
shape=(1, 30)>
        return %11

    %6 = subgraph(%5):

         %0 = 0                                 # ClearScalar<uint1>
         %1 = 255                               # ClearScalar<uint8>
         %2 = 0                                 # ClearScalar<uint1>
         %3 = 0.003919792729613519              # ClearScalar<float64>
         %4 = 1.0                               # ClearScalar<float64>
         %5 = 1.0                               # ClearScalar<float64>
         %6 = [0.11045581]                      # ClearTensor<float32,
shape=(1,)>
         %7 = 0.0003101464749011019             # ClearScalar<float64>
         %8 = [[30]]                            # ClearTensor<uint5,
shape=(1, 1)>
         %9 = 0                                 # ClearScalar<uint1>
        %10 = input                             #
EncryptedTensor<uint5, shape=(1, 1)>
        %11 = astype(%10, dtype=float32)        #
EncryptedTensor<float32, shape=(1, 1)>
        %12 = add(%11, %9)                      #
EncryptedTensor<float32, shape=(1, 1)>
        %13 = add(%12, %8)                      #
EncryptedTensor<float64, shape=(1, 1)>
        %14 = multiply(%7, %13)                 #
EncryptedTensor<float64, shape=(1, 1)>
        %15 = add(%14, %6)                      #
EncryptedTensor<float64, shape=(1, 1)>
        %16 = negative(%15)                     #
EncryptedTensor<float64, shape=(1, 1)>
        %17 = exp(%16)                          #
EncryptedTensor<float64, shape=(1, 1)>
        %18 = add(%5, %17)                      #
EncryptedTensor<float64, shape=(1, 1)>
        %19 = divide(%4, %18)                   #
EncryptedTensor<float64, shape=(1, 1)>
        %20 = divide(%19, %3)                   #
EncryptedTensor<float64, shape=(1, 1)>
        %21 = add(%20, %2)                      #
EncryptedTensor<float64, shape=(1, 1)>
        %22 = rint(%21)                         #
EncryptedTensor<float64, shape=(1, 1)>
        %23 = clip(%22, %0, %1)                 #
EncryptedTensor<float64, shape=(1, 1)>
        %24 = astype(%23, dtype=int_)           #
EncryptedTensor<uint1, shape=(1, 1)>
        return %24

**Comment by fd0r** [comment]
Which version do you have installed on your system?
Could you please make sure that you are using the latest version 1.3.0 of Concrete ML.

**Comment by berhankarabay** [comment]
No, I'm using 0.3.0. I've never thought of it, thank you. But how can I
update to the latest version? Update command didn't work on the terminal.

**Comment by RomanBredehoft** [comment]
Hello @berhankarabay , 
Running `pip install concrete-ml==1.3.0` in your terminal should update Concrete ML properly !

**Comment by berhankarabay** [comment]
Tried that, any idea why the following error occurred?

Collecting concrete-ml==1.3.0

  Using cached concrete_ml-1.3.0-py3-none-any.whl.metadata (12 kB)

Collecting boto3<2.0.0,>=1.23.5 (from concrete-ml==1.3.0)

  Using cached boto3-1.29.2-py3-none-any.whl.metadata (6.7 kB)

Collecting brevitas==0.8.0 (from concrete-ml==1.3.0)

  Using cached brevitas-0.8.0-py3-none-any.whl (357 kB)

Collecting certifi==2023.07.22 (from concrete-ml==1.3.0)

  Using cached certifi-2023.7.22-py3-none-any.whl.metadata (2.2 kB)

INFO: pip is looking at multiple versions of concrete-ml to determine which
version is compatible with other requirements. This could take a while.

ERROR: Could not find a version that satisfies the requirement
concrete-python==2.5.0-rc1 (from concrete-ml) (from versions: none)

ERROR: No matching distribution found for concrete-python==2.5.0-rc1

**Comment by berhankarabay** [comment]
Hi All,

I'm sorry for taking your time. However, I couldn't solve the issue after
many attempts.

Now I'm aware that I need to upgrade concrete-ml, however I couldn't
install 1.3.0. Please let me know if you have any suggestions on how to
proceed?

Best regards,

**Comment by berhankarabay** [comment]
Hello again, 

I wanted to summarize the issue to avoid long conversation here. I've tried to use concrete-ml through native installation, on a conda environment and a virtual environment with Python 3.8 and 3.9. When I downloaded concrete-ml, I got the version 0.3.0. I tested it with the [simple example](https://docs.zama.ai/concrete-ml) given in your website, but it didn't work.

When I tried to install version 1.3.0 using the following code: pip install concrete-ml==1.3.0,
I get the error below. I've tried downloading concrete-python to solve the issue, but unable to do it as well. 

Collecting concrete-ml==1.3.0
Using cached concrete_ml-1.3.0-py3-none-any.whl.metadata (12 kB)
Collecting boto3<2.0.0,>=1.23.5 (from concrete-ml==1.3.0)
Using cached boto3-1.29.2-py3-none-any.whl.metadata (6.7 kB)
Collecting brevitas==0.8.0 (from concrete-ml==1.3.0)
Using cached brevitas-0.8.0-py3-none-any.whl (357 kB)
Collecting certifi==2023.07.22 (from concrete-ml==1.3.0)
Using cached certifi-2023.7.22-py3-none-any.whl.metadata (2.2 kB)

INFO: pip is looking at multiple versions of concrete-ml to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement concrete-python==2.5.0-rc1 (from concrete-ml) (from versions: none)
ERROR: No matching distribution found for concrete-python==2.5.0-rc1


Can you help me with this? How can I use the 1.3.0 version of concrete-ml? In your website, you say Apple Silicon is only supported with the Docker option. However here it says Docker doesn't work as well? 

Best,

**Comment by berhankarabay** [comment]
Hello again, 

I solved the issue, it turns out that my python architecture was incorrect. I fixed it to executable arm64 and installed 1.3.0. Now, while I'm running the [simple example](https://docs.zama.ai/concrete-ml) given in your website, I received the following error. 

File "/Users/berhankarabay/Desktop/COVENTRY/dissertation/lib/python3.9/site-packages/concrete/compiler/library_support.py", line 170, in compile
    self.cpp().compile(

RuntimeError: Caught an unknown exception!

Do you have any suggestion? I have no clue how to fix this one. Best,
